internal/nvmeof/nvmeof.go

```
@@ -0,0 +1,424 @@
+limitations under the License.
+*/
+
+package nvmeof
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"syscall"
+
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+)
+
+// GatewayAddress holds the address and port of the NVMe-oF gateway.
+// This is used to connect to the gateway server.
+// It is also used to specify the address and port of listeners.
+type GatewayAddress struct {
+	Address string
+	Port    uint32
+}
+
+func (ga GatewayAddress) String() string {
+	return fmt.Sprintf("%s:%d", ga.Address, ga.Port)
+}
+
+// ListenerDetails holds the listener information for a subsystem.
+type ListenerDetails struct {
+	GatewayAddress
+	Hostname string
+}
+
+// Config holds gateway client configuration.
+type GatewayConfig = GatewayAddress
+
+// NVMeoFVolumeData holds the data required for an NVMe-oF volume.
+type NVMeoFVolumeData struct {
+	SubsystemNQN          string
+	NamespaceID           uint32
+	HostNQN               string
+	ListenerInfo          ListenerDetails
+	GatewayManagementInfo GatewayConfig
+}
+
+// GatewayClient wraps the gRPC client and connection details.
+type GatewayRpcClient struct {
+	config    *GatewayConfig
+	conn      *grpc.ClientConn
+	client    GatewayClient // protobuf client
+	maxSerial *big.Int      // Maximum allowed serial number for subsystems.
+}
+
+// TODO: Move this to a util package?
+const (
+	MaxAllowedSubsystemSerialNumber = 99999999999999
+	RandomNumberOffset              = 2 // skip reserved identifiers 0 and 1
+)
+
+// NewGatewayRpcClient creates a new gateway client.
+func NewGatewayRpcClient(config *GatewayConfig) (*GatewayRpcClient, error) {
+	client := &GatewayRpcClient{
+		config:    config,
+		maxSerial: big.NewInt(MaxAllowedSubsystemSerialNumber - RandomNumberOffset),
+	}
+
+	// Only connect if config is provided
+	if config != nil {
```

what is the use of returning an unconnected client? in all other functions we need to check if the client is empty or not before accessing it. wont it be better to return an error if config is empty?

---

`config` is never `nil` in the current code, so dropped this check.